### DISCUSSION TITLE
Double voting with clock drift

### DISCUSSION THREAD

**Discussion by rsolari** [pooltogether/multichain-gov-proposal-spec#4] [General]
I like how minimal the spec is. I expect that we might need to add more state transitions, though, if we want to prevent someone from voting multiple times on different chains with the same votes.

The general problem is that the root node only knows about vote totals from the branches. It does not know which tokens voted on each branch. Tokens are supposed to only exist in one place at one time, but messages do not travel instantly. Some potential solutions to clock drift:

1. Don't worry about it. Allow multiple voting if someone wants to frontrun the bridges.
2. Slow down delegations. Add a conservative – an hour? a day? – period between delegating a vote to an address and increasing the voting power of that address. To do that, we'd have to either require two on-chain actions from the user or keep track of delegation history
3. Slow down bridging. Somehow maintain the invariant that bridging tokens is always slower than bridging a proposal. I don't understand bridges well enough to know if this approach is reasonable.

**Comment by asselstine** [comment]
Yes! The spec is very minimal.  By keeping it strictly focused on proposal structure, I think it'll be easier to reach consensus. However, the question of voting is critically important. I'm wondering whether that should be part of this spec, or outlined separately. The diagram includes the "addVotes" call, but that call isn't mentioned anywhere else.

Intuitively, I think projects could have different ways of measuring voting power; like burning an NFT or something. Does it make sense to keep this spec as abstract as possible and create a new spec for an approach to voting, or mix the two?

Needs more thought.

Regarding the problem of clock drift itself, I agree that we have to solve it. Intuitively, it seems like we need more fuzziness with regard to voting power.  For example:

Let's say the proposal startTime = 10s
Clock drift can be +/- 2s

The startTime is determined by the GovernorRoot, so the local time for branches on other chains may be +/- 2s.  What we're essentially saying here is that **all chain clocks overlap somewhere between 8 and 12 seconds**. One may be 1 second behind, another 2 ahead.

This rule can be useful to us; since those time ranges over lap, then we can prevent double spending by requiring users to hold voting power *for the duration of the time range*! I think this style of voting more closely resembles the "lock your tokens to vote" approach, similar to veCRV.

**Comment by frangio** [comment]
Avoiding double voting is critical and I think the solution should be part of this spec.

It seems true that if we assume a given worst case drift, and we take a time range long enough to cover that drift, then all chain clocks should overlap somewhere in the range. If it's +/- 2s I think we probably want a range quite a bit longer than 4s for safety margin though.

To put it in other words, a multi-chain governance system would organize around "epochs" (timestamp modulo epoch length) designed to be long enough to guarantee that different chains' epochs have some overlap even though they have different clocks, and the voting period for a proposal would be defined in terms of start and end epochs. Importantly, votes shouldn't move between accounts in one epoch.

Locking votes is one way of implementing this behavior, but we can also explore "slow down delegations" as proposed by @rsolari initially above. Both things would work as long as they are organized around these epochs.

**Comment by asselstine** [comment]
The concept of a voting `epoch` is very interesting!

I created a PR for a first pass at a [locked votes approach](https://github.com/pooltogether/multichain-gov-proposal-spec/blob/voting/MultiChainVoting.md).  While it doesn't have epochs, the one rule I did realize is:

A users voting power should be the **minimum balance they held** between (`timestamp - range`, `timestamp`) where timestamp is the timestamp of the proposal.  This prevents double-voting because it ignores transfers out and in for that range. However, the PR essentially treats the block interval as the epoch, so the granularity requires votes to be locked. That sucks!

I really like this concept of epochs.  If the users voting power was **the minimum balance they held during the epoch** then I think that would enforce the same rule as above, and let wrapped voting tokens be fully liquid without any locking concerns.

I'm going to play around with this idea and update the PR.
